### ISSUE TITLE
smolvla model not working properly

### ISSUE THREAD

**Issue by yuyoujiang** [huggingface/lerobot#1239]
I trained an ACT with 30 episodes data, and the entire training process went smoothly without any issues. The testing process was also very smooth. However, when I switched the model to SmolVLA, the model failed to function properly during inference.

This is the dataset I used：
https://huggingface.co/datasets/youjiang97/so100_tt 

These are the commands I used for training and testing：
```bash
python lerobot/scripts/train.py   --dataset.repo_id=${HF_USER}/so100_tt   --policy.type=smolvla   --output_dir=outputs/train/smolvla_so100_0   --job_name=smolvla_so100_0   --policy.device=cuda 
```

```bash
python lerobot/scripts/control_robot.py   --robot.type=so100   --control.type=record   --control.fps=30   --control.single_task="Grasp a lego block and put it in the bin."   --control.repo_id=${HF_USER}/eval_smolvla_so100_tt   --control.tags='["tutorial"]'   --control.warmup_time_s=5   --control.episode_time_s=60   --control.reset_time_s=5   --control.num_episodes=10   --control.push_to_hub=true   --control.policy.path=outputs/train/smolvla_so100_1/checkpoints/last/pretrained_model
```

Here is the failed demo:
https://www.youtube.com/watch?v=kgVInbudhiw

Here is the training configuration file：
[train_config.json](https://github.com/user-attachments/files/20650017/train_config.json)

Any advice would be greatly appreciated.

**Comment by kasiv008** [comment]
Same here. I have ported lerobot to support my xarm u850. Trained ACT for 40 episodes and it achieves good grasping with repeatability. Trained smolVLA which went smoothly,  but the inference fails. It approaches the object to grasp but gets stuck in a loop as you can see in the video.

ACT inference:
[Act](https://youtube.com/shorts/DEtMyughkrE?feature=share)
SmolVLA inference:
[SmolVLA](https://youtube.com/shorts/qdtgxaju9YI?feature=share)

**Comment by danaaubakirova** [comment]
Hi there!

Thanks for sharing your dataset and the feedback on SmolVLA. I took a look, it's quite complex setup, and with only 30 episodes and varying object positions, it's tough for current robotics models to learn robustly.

I recommend checking out this dataset for reference:
🔗 [SVLA SO100 PickPlace](https://huggingface.co/spaces/lerobot/visualize_dataset?path=%2Flerobot%2Fsvla_so100_pickplace%2Fepisode_0)

In this dataset, we recorded 50 episodes across 5 distinct cube positions. For each position, we collected 10 episodes of pick-and-place interactions. This structure, repeating each variation several times, helped the model generalize better. We tried similar dataset with 25 episodes, and it was not enough leading to a bad performance. So, the data quality and quantity is definitely a key!

**Comment by ZhuYaoHui1998** [comment]
I collected 100 sets of data where a green block was placed into a fixed basket, with the green block randomly positioned. Two cameras were used. However, SmoLVLA still keeps shaking and looping head movements at the initial position after 20k training steps, even though the loss has dropped to 0.021. How can I fix this? Under the same dataset, ACT works fine.

**Comment by kasiv008** [comment]
A small update. As suggested by @danaaubakirova, I increased the number of episodes to 80 and got SmolVLA to working!
Thanks.

**Comment by ZhuYaoHui1998** [comment]
But I record 100 datasets, i am sad

**Comment by chirag1701** [comment]
> But I record 100 datasets, i am sad

I faced the same problem with 50 Episodes on pick_place (https://huggingface.co/chirag1701/so101_pickplace_B_smolvla_cwnu)
I will now try with 90 episodes. Any update from your side? Did you solve the problem?

**Comment by ZhuYaoHui1998** [comment]
no, its still not work.

**Comment by chirag1701** [comment]
I have had success with smolvla with making many episodes for the same position. And make for multiple positions. Don't make just 1 epsiode for 1 position

**Comment by Liu-Xiaoyan97** [comment]
> 你好！
> 
> 感谢您分享您的数据集和对SmolVLA的反馈。我看了看，它的设置相当复杂，只有30集和不同的物体位置，目前的机器人模型很难稳健地学习。
> 
> 我建议查看此数据集以供参考：[🔗SVLA SO100 PickPlace](https://huggingface.co/spaces/lerobot/visualize_dataset?path=%2Flerobot%2Fsvla_so100_pickplace%2Fepisode_0)
> 
> 在这个数据集中，我们在5个不同的立方体位置记录了50集。对于每个职位，我们收集了10集的挑选和放置互动。这种结构，多次重复每个变化，有助于模型更好地推广。我们尝试了25集的类似数据集，这还不够，导致表现不佳。所以，数据质量和数量绝对是关键！

> Hi there!
> 
> Thanks for sharing your dataset and the feedback on SmolVLA. I took a look, it's quite complex setup, and with only 30 episodes and varying object positions, it's tough for current robotics models to learn robustly.
> 
> I recommend checking out this dataset for reference: 🔗 [SVLA SO100 PickPlace](https://huggingface.co/spaces/lerobot/visualize_dataset?path=%2Flerobot%2Fsvla_so100_pickplace%2Fepisode_0)
> 
> In this dataset, we recorded 50 episodes across 5 distinct cube positions. For each position, we collected 10 episodes of pick-and-place interactions. This structure, repeating each variation several times, helped the model generalize better. We tried similar dataset with 25 episodes, and it was not enough leading to a bad performance. So, the data quality and quantity is definitely a key!

I recorded 77 episodes across 5 distinct positions. But it still worked fail. 
[Xiaoyan97/orange_block_pickplace](https://huggingface.co/datasets/Xiaoyan97/orange_block_pickplace/tree/main)

**Comment by Liu-Xiaoyan97** [comment]
> no, its still not work.

did you solve the problem?

**Comment by chirag1701** [comment]
@Liu-Xiaoyan97 In your case, i think the problem is with the camera positions you have taken. There should be a full context view for the robot to see the whole broad setup. Thats why its suggested to keep one "up" camera for the full view and the other "gripper" for precision in picking the blocks. But, both of the cameras in your case almost give similar views and hence the model is not able to pinpoint the location of the blocks. Let me know if you face any problem :)

**Comment by chirag1701** [comment]
[chirag1701/new0_green_final](https://huggingface.co/datasets/chirag1701/new0_green_final)
refer to this dataset for reference

**Comment by danaaubakirova** [comment]
Thanks everyone for sharing your datasets and observations. It's been really helpful to understand the different cases where SmolVLA struggles.

**TL;DR:**
SmolVLA is more sensitive than ACT to data quality, episode length, and camera setup. If actions are too fast, or visual coverage is limited (e.g., no top view in [Xiaoyan97/orange_block_pickplace](https://huggingface.co/datasets/Xiaoyan97/orange_block_pickplace)), the model tends to fail during inference, e.g. getting stuck in loops or showing shaky behavior. This usually indicates failure to properly learn during training.

**Why short actions or fast episodes can cause failure?**
From our experience, very short episodes with quick actions often result in poor learning.
When the robot moves too quickly, the model receives fewer frames to understand how actions affect the state. This weakens the ability to learn meaningful transitions. Attaching some visuals examples from what you have shared here: 

Good example:
[Episode 99](https://huggingface.co/spaces/lerobot/visualize_dataset?path=%2Fchirag1701%2Fnew0_green_final%2Fepisode_99) – ~16 seconds (~480 frames at 30 FPS)
→ Smooth, well-paced motion with clear progression. Easier for the model to learn.

Less effective example:
[Episode 0](https://huggingface.co/spaces/lerobot/visualize_dataset?path=%2FXiaoyan97%2Forange_block_pickplace%2Fepisode_0) – ~10 seconds (~300 frames)
→ Too short and fast, with abrupt joint changes. Harder for the model to extract motion cues.

**Recommendations for better SmolVLA performance**

1. Record actions slowly and smoothly. Avoid fast or jerky motions.

2. Don’t collect just one episode per object position. Use 5–10 episodes per position across unique positions you want to have.

3. Use a top-view camera that captures the full workspace, so the model sees the full spatial context (arm, object layout, goal area). Use a gripper-view camera for precision. If both cameras are too similar (e.g., both zoomed-in), the model may fail to localize objects reliably.

I know this may sound like a lot of setup rules, but it's all part of building more robust VLA models. The better our datasets, the better these models will generalize and perform :)

If your dataset is already recorded:

- Consider filtering out very short or abrupt episodes
- Make sure action and observation signals are properly aligned
- Watch for issues like delayed or inconsistent gripper signals

More data recording tips are available here:
🔗 https://huggingface.co/blog/lerobot-datasets

Feel free to share specific episodes or demos if you'd like feedback or help debugging.

**Comment by chiragr-jarm** [comment]
Hey everyone, thank you for the good examples and explanations. I have about a 100 episodes recorded for a pick and place task, but still don't know why it doesn't work. @danaaubakirova Could please provide some feedback for this dataset: pooja420/so101_color_test_combined

I would like to see what I am missing

**Comment by chirag1701** [comment]
1. There are a few very bad episodes and it's a good practice not to add those episodes in the training set. I think a very bad episode can spoil other 5 good episodes.
2. The task is a bit tricky for first time. First , get the model working on simple pick and place task. Also, start with just 5-6 fixed positons and keep switching between those while recording. For your task, the training dataset has many instances where the teleoperater is confused about which cube to pick. Also, it would be better to keep the cubes a little more apart.
3. I think top view is really necessary for the model to get good at identifying positions. Prefer top view instead of side view.
4. KEEP THE Top/side cameras fully anchored to their positions. Any small change could lead to a fixed offset while testing.

You can share a demonstration of testing and i would be better able to identify the reason. Also, if the robot is crashing or something like that, check if all the motors are alright and calibration is done properly

**Comment by smithasb4** [comment]
Hi! I am experiencing a similar issue with spatial accuracy using SmolVLA on an SO101 arm.

**Setup**:

**Dataset**: 100 episodes (20k steps for training) of a 'Press Red Button' task.(recorded 10 episodes for 10 different positions) ([Link to dataset](https://huggingface.co/spaces/lerobot/visualize_dataset?path=%2Fsmithasbellavi%2Fdataset_PressLitRedButton%2Fepisode_0))

**Cameras**: 3 views (Top-front, Top-side, and Wrist).

**Issue**
During inference, the model appears to locate the center of the board rather than the specific button. However, if I manually move the board so the button enters the wrist camera's FOV, the model regains tracking and successfully completes the task (pressing and returning to home).

How can I resolve this issue? Are the number of episodes insufficient?

**Comment by nicholas-maselli** [comment]
If using SmolVLA @smithasb4, 100 episodes (what is the batch size?) with 3 cameras, 20k steps of training is likely too low.

This does depend on camera resolution / loss etc though. SmolVLA is larger then ACT, If you take more data and train longer you should be able to get the arms to do this task successfully

**Comment by smithasb4** [comment]
> what is the batch size? @nicholas-maselli 
32

Previous to this, I trained it with 50k steps and a batch size of 64. This model was worse. It tried to press the button somewhere in the air above the board.

I even trained with 220 episodes with 35k steps and 32 batch size. This was not any better either. And there was no understanding of whether the light is on or not.

**Comment by nicholas-maselli** [comment]
That's a bit surprising @smithasb4 these numbers should be enough. My recommendation here is train the dataset with ACT to see the performance difference, and try to look over your data (maybe even train a SARM model with the data) to determine quality of the data as SmolVLA requires higher quality data compared to ACT as per the comment above and personal experience

**Comment by smithasb4** [comment]
> That's a bit surprising [@smithasb4](https://github.com/smithasb4) these numbers should be enough. My recommendation here is train the dataset with ACT to see the performance difference, and try to look over your data (maybe even train a SARM model with the data) to determine quality of the data as SmolVLA requires higher quality data compared to ACT as per the comment above and personal experience

I trained an ACT model with the same dataset, for 100k steps and 32 batch size. The result is the same. There is still an offset from the button, but if the wrist camera can see it, it can re-position itself to press the button. Is 100 episodes too less for the model?

**Comment by devika-dudo** [comment]
> I collected 100 sets of data where a green block was placed into a fixed basket, with the green block randomly positioned. Two cameras were used. However, SmoLVLA still keeps shaking and looping head movements at the initial position after 20k training steps, even though the loss has dropped to 0.021. How can I fix this? Under the same dataset, ACT works fine.
> 

Hii
Even the smolVLA I trained seems stuck during inference...
ACT was way better

**Comment by devika-dudo** [comment]
> I have had success with smolvla with making many episodes for the same position. And make for multiple positions. Don't make just 1 epsiode for 1 position 

Actually....I was trying to get a generalized model working....is it possible to achieve that??

**Comment by devika-dudo** [comment]
> Thanks everyone for sharing your datasets and observations. It's been really helpful to understand the different cases where SmolVLA struggles.
> 
> **TL;DR:**
> SmolVLA is more sensitive than ACT to data quality, episode length, and camera setup. If actions are too fast, or visual coverage is limited (e.g., no top view in [Xiaoyan97/orange_block_pickplace](https://huggingface.co/datasets/Xiaoyan97/orange_block_pickplace)), the model tends to fail during inference, e.g. getting stuck in loops or showing shaky behavior. This usually indicates failure to properly learn during training.
> 
> **Why short actions or fast episodes can cause failure?**
> From our experience, very short episodes with quick actions often result in poor learning.
> When the robot moves too quickly, the model receives fewer frames to understand how actions affect the state. This weakens the ability to learn meaningful transitions. Attaching some visuals examples from what you have shared here: 
> 
> Good example:
> [Episode 99](https://huggingface.co/spaces/lerobot/visualize_dataset?path=%2Fchirag1701%2Fnew0_green_final%2Fepisode_99) – ~16 seconds (~480 frames at 30 FPS)
> → Smooth, well-paced motion with clear progression. Easier for the model to learn.
> 
> Less effective example:
> [Episode 0](https://huggingface.co/spaces/lerobot/visualize_dataset?path=%2FXiaoyan97%2Forange_block_pickplace%2Fepisode_0) – ~10 seconds (~300 frames)
> → Too short and fast, with abrupt joint changes. Harder for the model to extract motion cues.
> 
> **Recommendations for better SmolVLA performance**
> 
> 1. Record actions slowly and smoothly. Avoid fast or jerky motions.
> 
> 2. Don’t collect just one episode per object position. Use 5–10 episodes per position across unique positions you want to have.
> 
> 3. Use a top-view camera that captures the full workspace, so the model sees the full spatial context (arm, object layout, goal area). Use a gripper-view camera for precision. If both cameras are too similar (e.g., both zoomed-in), the model may fail to localize objects reliably.
> 
> I know this may sound like a lot of setup rules, but it's all part of building more robust VLA models. The better our datasets, the better these models will generalize and perform :)
> 
> If your dataset is already recorded:
> 
> - Consider filtering out very short or abrupt episodes
> - Make sure action and observation signals are properly aligned
> - Watch for issues like delayed or inconsistent gripper signals
> 
> More data recording tips are available here:
> 🔗 https://huggingface.co/blog/lerobot-datasets
> 
> Feel free to share specific episodes or demos if you'd like feedback or help debugging.

Hii...if we train only for 5 positions...will it only pick from those positions during inference too??

**Comment by chirag1701** [comment]
> > I have had success with smolvla with making many episodes for the same position. And make for multiple positions. Don't make just 1 epsiode for 1 position
> 
> Actually....I was trying to get a generalized model working....is it possible to achieve that??

Sorry, smolvla is not that capable right now

**Comment by devika-dudo** [comment]
> > > I have had success with smolvla with making many episodes for the same position. And make for multiple positions. Don't make just 1 epsiode for 1 position
> > 
> > 
> > Actually....I was trying to get a generalized model working....is it possible to achieve that??
> 
> Sorry, smolvla is not that capable right now

for teaching just one task to a bimanual mobile robot --- do we have a generalized model yet....other than smolVLA  
have you found any good results with any other models??
